src/cmd/builtin/dmesg.c

```diff
@@ -26,122 +26,66 @@ static struct optparse_option dmesg_opts[] = {
-        if (!(f = dmesg_rpc (h, follow)))
-            goto done;
-        while (!eof) {
-            if (dmesg_rpc_get (f, fun, arg) < 0) {
-                if (errno != ENODATA) {
-                    flux_future_destroy (f);
-                    goto done;
-                }
-                eof = true;
-            }
-            flux_future_reset (f);
-        }
-        flux_future_destroy (f);
-    }
-    if ((flags & DMESG_CLEAR)) {
-        if (dmesg_clear (h) < 0)
-            goto done;
-    }
-    rc = 0;
-done:
-    return rc;
-}
-
-void dmesg_fprint (const char *buf, int len, void *arg)
-{
-    FILE *f = arg;
     struct stdlog_header hdr;
     const char *msg;
     int msglen, severity;
     uint32_t nodeid;
 
-    if (f) {
-        if (stdlog_decode (buf, len, &hdr, NULL, NULL, &msg, &msglen) < 0)
-            fprintf (f, "%.*s\n", len, buf);
-        else {
-            nodeid = strtoul (hdr.hostname, NULL, 10);
-            severity = STDLOG_SEVERITY (hdr.pri);
-            fprintf (f, "%s %s.%s[%" PRIu32 "]: %.*s\n",
-                     hdr.timestamp,
-                     hdr.appname,
-                     stdlog_severity_to_string (severity),
-                     nodeid,
-                     msglen, msg);
-        }
-        fflush (f);
+    if (stdlog_decode (buf, len, &hdr, NULL, NULL, &msg, &msglen) < 0)
+        fprintf (f, "%.*s\n", len, buf);
+    else {
+        nodeid = strtoul (hdr.hostname, NULL, 10);
+        severity = STDLOG_SEVERITY (hdr.pri);
+        fprintf (f, "%s %s.%s[%" PRIu32 "]: %.*s\n",
+                 hdr.timestamp,
+                 hdr.appname,
+                 stdlog_severity_to_string (severity),
+                 nodeid,
+                 msglen, msg);
     }
+    fflush (f);
 }
 
 static int cmd_dmesg (optparse_t *p, int ac, char *av[])
 {
     int n;
     flux_t *h;
-    int flags = 0;
-    flux_log_f print_cb = dmesg_fprint;
 
     if ((n = optparse_option_index (p)) != ac)
         log_msg_exit ("flux-dmesg accepts no free arguments");
 
     h = builtin_get_flux_handle (p);
-    if (optparse_hasopt (p, "read-clear") || optparse_hasopt (p, "clear"))
-        flags |= DMESG_CLEAR;
-    if (optparse_hasopt (p, "clear"))
-        print_cb = NULL;
-    if (optparse_hasopt (p, "follow"))
-        flags |= DMESG_FOLLOW;
-    if (dmesg (h, flags, print_cb, stdout) < 0)
-        log_err_exit ("log.dmesg");
+
+    if (!optparse_hasopt (p, "clear")) {
+        flux_future_t *f;
+        const char *buf;
+
+        if (!(f = flux_rpc_pack (h,
+                                 "log.dmesg",
+                                 FLUX_NODEID_ANY,
+                                 FLUX_RPC_STREAMING,
+                                 "{s:b}",
+                                 "follow", optparse_hasopt (p, "follow"))))
+            log_err_exit ("error sending log.dmesg request");
+        while (flux_rpc_get (f, &buf) == 0) {
+            dmesg_fprint (stdout, buf, strlen (buf));
```

Comment:
AFAICT we only output to stdout and no other file streams are possible?  We could save ourselves the extra parameter to `dmesg_fprintf()` and remove the `FILE *f` parameter.  Thus we could just use `printf` in `dmesg_fprintf()` ... and I guess `dmesg_fprintf()` could be renamed into `dmesg_output()` or something.